Add semantic hashing

abhi3700 opened the pull request:
## Description

This PR is about adding demo of Buckets of semantic hashes for given set of search queries.

Based on this [Notion doc](https://www.notion.so/subspacelabs/Semantic-Hashing-Demo-38297cb7da594dcfb96393a3c491a936).

abhi3700 committed:
Add semantic hashing including buckets

abhi3700 committed:
Added the casting into np.asarray

abhi3700 committed:
Fix linting issue

abhi3700 committed:
Add a knowledge base & compare bucket's key with hash of search query

abhi3700 committed:
Fmt done

abhi3700 committed:
Add hamming distance to compare query with knowledge base

0xjeremyfrank commented on src/semantic_hashing_demo/main.py:
```diff
@@ -0,0 +1,143 @@
+        str: A binary string of length nbits.
+    """
+    # create a set of 4 hyperplanes (normal ⟂), with 2 dimensions
+    plane_norms = np.random.rand(int(nbits), len(v)) - 0.5
```
Will this create different random planes every time? If so, this will make the hashes non-comparable.

abhi3700 replied:
Fixed this with seeded hyperplanes now. https://github.com/abhi3700/semantic-hashing-demo/commit/c42872b2224b2ef3f324842c322d237be77ef90d

abhi3700 committed:
Add index of the bucket that the given text belongs to

abhi3700 committed:
Add code to load real dataset from CSV

- add polars package instead of pandas because of its speed.
- show the bucket map key which the given review text belongs to.
- format the terminal display for proper understanding.

abhi3700 committed:
Add seeded hyperplanes using RandomState

- 'subspace' is used as default seed.
- code verified with small datasets (due to time) that the given query does belong to the expected bucket which was formed with the bucket training reviews dataset.

abhi3700 committed:
minor change

abhi3700 committed:
Add benchmark results with 10 samples, 4/8 hyperplanes with the default seed

abhi3700 committed:
Format execution time unit

abhi3700 committed:
Update readme with pre-requisite for running the repo

abhi3700 committed:
Fix a major bug to get deterministic buckets in each run

- Getting deterministic buckets each time with same sample-size & seeded hyperplanes.
- seed changed from string to int.
- removed 'hash' fn as it throws different values for each run.
- verified deterministic buckets with multiple runs

abhi3700 committed:
Fmt done, Corrected linting issue

- renamed input to config as it is more appropriate as used here.
- importing from module like semant....config fixed the linting issue found during 'huak lint'.

abhi3700 committed:
Add manual algorithm tests for 3 search query types: same, modified, generated

abhi3700 committed:
Add conclusion for manual testing

0xjeremyfrank commented on data/README.md:
```diff
@@ -0,0 +1 @@
+Download the [data](https://www.kaggle.com/datasets/snap/amazon-fine-food-reviews) from kaggle. It's too big (~256 MB) to upload to github.
```
Since we are not using the whole dataset you should be able to upload the whole file.

abhi3700 replied:
At this point not sure, which reviews would be used especially in the case of testing. In some cases, I am picking any random data out of the samples (20/40) fed.
So, it would be better to download manually from the given link.

0xjeremyfrank replied:
I think you can still add the file as it is used now and note that it comes from a bigger dataset with the link. This will make setup much better.

abhi3700 replied:
Ok I will trim it & upload then.

abhi3700 replied:
Added in this commit: https://github.com/abhi3700/semantic-hashing-demo/commit/b2245654fa144366eb2e2ef9e6232497a864759a

0xjeremyfrank commented on README.md:
```diff
@@ -0,0 +1,51 @@
+
+1. `huak` is installed following the [guide](https://github.com/cnpryer/huak/blob/master/docs/user_guide.md#installation).
+2. Download the data file as per [README](./data/README.md).
+
```
you need to `huak install` as well.

abhi3700 replied:
Instead you need to `huak build`

0xjeremyfrank replied:
that didn't work for me 🤔

abhi3700 replied:
What is the error?

0xjeremyfrank commented on tests/README.md:
```diff
@@ -0,0 +1,122 @@
+Had to also consider openai large embedding model as well. In some cases, tests passed as in the query text fell into the expected bucket.
+At this point it’s a bit ambiguous to jump to a conclusion. There are 2 main observations as of now:
+
+1. In cases where the semantic hash is different for the original & query texts, embedding model needs to be replaced.
```
I don't think this is true, we lose a lot of information from the original vector embedding to the hash. I don't think it will make a huge difference if you use an improved embedding model.

abhi3700 replied:
Commented (Removed) this thought as of now.

0xjeremyfrank commented on tests/README.md:
```diff
@@ -0,0 +1,122 @@
+At this point it’s a bit ambiguous to jump to a conclusion. There are 2 main observations as of now:
+
+1. In cases where the semantic hash is different for the original & query texts, embedding model needs to be replaced.
+2. In cases where the semantic hash is same for the original & query texts, hamming distance vector is to blamed for. For instance, with [2, 1, 1, 3, 4] hamming distance vector, the traversal if done from left, the query text is supposed to go into bucket at index-1 as it is the shortest distance so far. But, the original text is inside the bucket at index-2.
```
I don't follow #2?

abhi3700 replied:
Updated as per latest commit

0xjeremyfrank reviewed (commented):
Some thoughts, we discussed some of this already

- skip the exact text tests
- shrink your samples to 20
- I would increase nbits to 8
- I would have three categories of "texts", with 3 examples of each
  - Slightly modified
  - AI generated - instead of AI generated, let's call this "similar content" or something AI generated may be confusing in this context.
  - Completely unrelated

The expectations are that slightly modified would have low Hamming distance values, similar content will probably have higher values (not sure actually) and completely unrelated should have pretty high and random hamming values. Let me know if you have questions.

abhi3700 committed:
Add more tests related to more hyperplanes: 8 & less samples: 20

abhi3700 committed:
Added more tests

abhi3700 committed:
Structured the test remarks for type-2, 3

abhi3700 committed:
Add conclusion for type-3 i.e. similar text aka AI-generated text

abhi3700 committed:
Add tests for Unrelated text

abhi3700 commented:
@jfrank-summit
Added the required tests.
Still inconclusive with certainty.

abhi3700 committed:
Update some remarks

0xjeremyfrank commented on src/semantic_hashing_demo/main.py:
```diff
@@ -0,0 +1,183 @@
+    Returns:
+        str: A binary string of length nbits.
+    """
+    # Create a RandomState instance with the seed
+    # rng = RandomState(seed_int)
+    rng = RandomState(seed)
+
+    # Generate hyperplanes using the seeded random number generator
+    plane_norms = rng.rand(int(nbits), len(v)) - 0.5
```
This code should not be run every time, it is highly inefficient. Maybe turn this into a class and create the random planes on instantiation.

abhi3700 replied:
Fixed. https://github.com/abhi3700/semantic-hashing-demo/commit/9d048e12840c73d309703f115955fc942dea2377

abhi3700 committed:
Improve code efficiency

- Take out the plane_norms from 'hash_vector' to avoid running in loops for each review text

abhi3700 committed:
data added
